File: BDA_backend/src/main/java/com/example/bda/Services/ZonaService.java
package com.example.bda.Services;

import com.example.bda.Model.zonas_urbanas;
import com.example.bda.Repository.ZonaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class ZonaService {

    @Autowired
    private ZonaRepository zonaRepository;

    // CREATE
    public zonas_urbanas create(zonas_urbanas zona) {
        int id = zonaRepository.create(zona);
        zona.setId_zona(id);
        return zona;
    }

    // READ - Obtener todos
    public List<zonas_urbanas> getAll() {
        return zonaRepository.findAll();
    }

    // READ - Obtener con paginación
    public Map<String, Object> getAllPaginated(int page, int size) {
        List<zonas_urbanas> zonas = zonaRepository.findAllPaginated(page, size);
        int total = zonaRepository.count();
        
        Map<String, Object> response = new HashMap<>();
        response.put("zonas", zonas);
        response.put("currentPage", page);
        response.put("totalItems", total);
        response.put("totalPages", (int) Math.ceil((double) total / size));
        return response;
    }

    // READ - Obtener por ID
    public zonas_urbanas getById(int id) {
        return zonaRepository.findById(id);
    }

    // UPDATE
    public zonas_urbanas update(int id, zonas_urbanas zona) {
        int rowsAffected = zonaRepository.update(id, zona);
        if (rowsAffected > 0) {
            zona.setId_zona(id);
            return zona;
        }
        return null;
    }

    // DELETE
    public boolean delete(int id) {
        return zonaRepository.delete(id) > 0;
    }

    // QUERIES SIMPLES
    public List<zonas_urbanas> getByTipo(String tipoZona) {
        return zonaRepository.findByTipo(tipoZona);
    }

    public List<zonas_urbanas> getByNombre(String nombre) {
        return zonaRepository.findByNombre(nombre);
    }

    public List<zonas_urbanas> getByAreaRange(float minArea, float maxArea) {
        return zonaRepository.findByAreaRange(minArea, maxArea);
    }

    // MÉTODO EXISTENTE
    public List<Map<String, Object>> obtenerZonasSinPlanificacion() {
        return zonaRepository.obtenerZonasSinPlanificacion();
    }
}

File: BDA_frontend/src/services/ZonaService.js
import api from './api';

const RESOURCE_URL = '/api/zona';

class ZonaService {
  // CRUD básico
  getAll() {
    return api.get(RESOURCE_URL);
  }

  getAllPaginated(page = 0, size = 10) {
    return api.get(`${RESOURCE_URL}/paginated`, { params: { page, size } });
  }

  getById(id) {
    return api.get(`${RESOURCE_URL}/${id}`);
  }

  create(zona) {
    return api.post(RESOURCE_URL, zona);
  }

  update(id, zona) {
    return api.put(`${RESOURCE_URL}/${id}`, zona);
  }

  delete(id) {
    return api.delete(`${RESOURCE_URL}/${id}`);
  }

  // Queries simples
  getByTipo(tipoZona) {
    return api.get(`${RESOURCE_URL}/tipo/${tipoZona}`);
  }

  getByNombre(nombre) {
    return api.get(`${RESOURCE_URL}/nombre`, { params: { nombre } });
  }

  getByAreaRange(minArea, maxArea) {
    return api.get(`${RESOURCE_URL}/area`, { params: { minArea, maxArea } });
  }

  // Método existente
  getSinPlanificacion() {
    return api.get(`${RESOURCE_URL}/sin-planificacion`);
  }

  // Alias en español
  obtenerTodos() {
    return this.getAll();
  }

  obtenerPaginado(page, size) {
    return this.getAllPaginated(page, size);
  }

  obtenerPorId(id) {
    return this.getById(id);
  }

  crear(zona) {
    return this.create(zona);
  }

  actualizar(id, zona) {
    return this.update(id, zona);
  }

  eliminar(id) {
    return this.delete(id);
  }
}

export default new ZonaService();
